### ISSUE TITLE
automounting of devicemappers fails with udisks2-backend

### ISSUE THREAD

**Issue by cherti** [coldfix/udiskie#71]
### background-setup:
external harddrive, luks-encrypted
-> udev-rule to call cryptsetup with a keyfile to open the harddrive automatically

### problem description:
With udisks1 as a backend, udiskie then mounts the devicemapper that appears as a consequence of the udev-rule doing its work. The device itself (containing the LUKS-volume) gets correctly recognized as not-to-be-handled.

However, when using udisks2 as a backend, udiskie ignores the upcoming devicemapper and asks for a password to unlock the harddrive (although it is already unlocked, the devicemapper is existing already, according to udiskie's debug-output below).

This behavior can be observed in version 1.2 with udisks2 as a default but in version 1.1.3 as well by explicitly using udisks2 as a backend. (earlier versions not tested, can be done if helpful).

## output (with -v) of udiskie 1.2.0:
Output generated by a running udiskie when plugging in a harddrive of the sort in question.
(udiskie command used: `udiskie -v -t -n -a [-(1|2)]`)
### For udisks1 (-1):
```
DEBUG [2015-07-29 02:07:57,200] udiskie.config: IgnoreDevice(match={'is_external': False}, value=True) used for /org/freedesktop/UDisks/devices/dm_2d1
DEBUG [2015-07-29 02:07:57,359] udiskie.config: IgnoreDevice(match={'is_external': False}, value=True) used for /org/freedesktop/UDisks/devices/dm_2d1
DEBUG [2015-07-29 02:07:57,359] udiskie.mount: mounting /org/freedesktop/UDisks/devices/dm_2d1 with {'fstype': 'ext4', 'options': None}
INFO [2015-07-29 02:07:57,514] udiskie.mount: mounted /org/freedesktop/UDisks/devices/dm_2d1 on /media/Epsilon
INFO [2015-07-29 02:07:57,551] udiskie.mount: not unlocking /org/freedesktop/UDisks/devices/sde: already unlocked
```

### for udisks2 (-2):
```
DEBUG [2015-07-29 02:00:51,724] udiskie.udisks2: +++ device_added: /org/freedesktop/UDisks2/block_devices/dm_2d1
DEBUG [2015-07-29 02:00:51,729] udiskie.config: IgnoreDevice(match={'is_external': False}, value=True) used for /org/freedesktop/UDisks2/block_devices/dm_2d1
DEBUG [2015-07-29 02:00:51,732] udiskie.udisks2: +++ device_added: /org/freedesktop/UDisks2/drives/WDC_WD30EURS_73TLHY0_WD_WMC4N0270075
DEBUG [2015-07-29 02:00:51,732] udiskie.config: IgnoreDevice(match={'is_block': False}, value=True) used for /org/freedesktop/UDisks2/drives/WDC_WD30EURS_73TLHY0_WD_WMC4N0270075
DEBUG [2015-07-29 02:00:51,741] udiskie.udisks2: +++ device_added: /org/freedesktop/UDisks2/block_devices/sde
```

**Comment by coldfix** [comment]
The debug output looks to me as if this is a problem with the `is_external` heuristic (for both versions of UDisks).

Try adding an explicit ignore rule, e.g. using the UUID to identify the devices:

```yaml
# ~/.config/udiskie/config.yml
ignore_device:
- id_uuid: "UUID of encrypted device"
  ignore: true
- id_uuid: "UUID of mapper device"
  ignore: false
```

(see also man page for further fields)

If this is indeed the problem, it would probably be nice to handle this more automatically, but I currently don't see any nice way of detecting this inside udiskie.

**Comment by coldfix** [comment]
By the way, I'm currently waiting for a patch to UDisks to be applied with which you could configure udiskie to unlock the device with a specified keyfile, which - I guess - would also solve this problem. Hope this gets included in the next 1-3 months or so.

**Comment by cherti** [comment]
Thanks for the proposal!

It does seem to work, unfortunately I end up not being authorized with udisks2:
```
[…]
DEBUG [2015-07-29 18:56:42,011] udiskie.mount: mounting /org/freedesktop/UDisks2/block_devices/dm_2d2 with {'options': None, 'fstype': 'ext4'}
ERROR [2015-07-29 18:56:42,019] udiskie.mount: failed to mount /org/freedesktop/UDisks2/block_devices/dm_2d2: GDBus.Error:org.freedesktop.UDisks2.Error.NotAuthorizedCanObtain: Not authorized to perform operation
[…]
```

**Comment by coldfix** [comment]
Oh, right makes sense. I guess the way to go is to create an udev rule that explicitly sets the "system" property of the device, as described [here](https://wiki.archlinux.org/index.php/Udev#Some_devices.2C_that_should_be_treated_as_removable.2C_are_not). This should also make the second config entry unnecessary.

If this doesn't work, I see two more rather unpleasant alternatives:

- [edit authorizations](auth_self) and set the permission for the "filesystem-mount-system" action to `AUTH_SELF` or `AUTH_SELF_KEEP` (see also [here](http://www.freedesktop.org/software/polkit/docs/latest/polkit.8.html#polkit-rules-polkit)). Of course this requires you to enter the password when mounting, which I believe is what you wanted to avoid by using keyfiles in the first place.. (I wouldn't recommend `YES` as this could be used to unmount your system devices from user seat)
- wait for the aforementioned patch to be applied and then configure udiskie to mount the device using keyfiles. In the meantime you could try to use an udev rule to mount your device.

**Comment by coldfix** [comment]
Let me know if this works, I should really add this to the README/wiki.

**Comment by cherti** [comment]
nope, setting UDISKS_SYSTEM and UDISKS_AUTO manually via udev doesn't help, unfortunately.

Both variables are set afterwards (according to udevadm), would you suspect that the problem is within udisks or rather within polkit? I'd like to try to get that sorted out upstream (I have the same issue with udisksctl, therefore I can confirm that this is no issue of udiskie).

There is one just-a-little unpleasant alternative I'll use for the moment: resorting back to udisks1. ;)
Of course that's no long term fix.

**Comment by coldfix** [comment]
Oh, that's a pitty. Did you set these flags on the encrypted device or on the mapper device (should be on the mapper device)?

I'd suspect the problem is with udisks. You can modify your polkit rule to check which permissions are queried, as described in the second half of [this post](https://github.com/coldfix/udiskie/issues/62#issuecomment-95555307).

Sorry, I should probably have tried to reproduce your issue on my machine to debug it on my own^^ but I was too lazy.

**Comment by cherti** [comment]
neither setting it on the block nor on the mapper helps.

No prob, I have the problematic setup here, therefore I can easily test with it. ;)

What I found out though is that with udisksctl mounting as a user is possible, as long as I provide my user's password. without entering a password: same error.

**Comment by coldfix** [comment]
Okay, FYI, I have now created a test device here and got it working. Hope the same solution applies for you too.

**tl;dr**: Make sure the ```filesystem-mount-other-seat`` permission is granted and the udev rule has a high prefix number and reload the rule via `udevadm control --reload`.


In detail, my setup goes is follows:

```bash
dd if=/dev/urandom of=/home/key count=4 bs=1024
cryptsetup luksFormat /dev/sdb1 /home/key
cryptsetup luksOpen /dev/sdb1 foo --key-file /home/key
mkfs.vfat /dev/mapper/foo -n STICKY

blkid -s UUID -o value /dev/sdb1
c0186b9a-5181-403f-a1fe-bcf101dd7d6a

blkid -s UUID -o value /dev/mapper/foo
D434-4277

cryptsetup luksClose foo
```

Udev rule `/etc/udev/rules.d/99-foobar.rules`. Note, chosing a higher prefix seems to make a difference (didn't work with 50-, don't know why):

```
ACTION=="add", ENV{ID_FS_UUID}=="c0186b9a-5181-403f-a1fe-bcf101dd7d6a", RUN+="/usr/bin/cryptsetup luksOpen /dev/%k foo --key-file /home/key"

ENV{ID_FS_UUID}=="D434-4277", ENV{UDISKS_AUTO}="1", ENV{UDISKS_SYSTEM}="0"
```

Udiskie config in `~/.config/udiskie/config.yml`:

```yaml
ignore_device:
  - id_uuid: c0186b9a-5181-403f-a1fe-bcf101dd7d6a
```

Polkit permissions in `/etc/polkit-1/rules.d/50-udiskie.rules`:

```javascript
polkit.addRule(function(action, subject) {

  // var prefix = "org.freedesktop.udisks";
  // if (action.id.slice(0, prefix.length) == prefix)
  //   polkit.log(action.id);

  var YES = polkit.Result.YES;
  var permission = {
    // only required for udisks1:
    "org.freedesktop.udisks.filesystem-mount": YES,
    "org.freedesktop.udisks.luks-unlock": YES,
    "org.freedesktop.udisks.drive-eject": YES,
    "org.freedesktop.udisks.drive-detach": YES,
    // only required for udisks2:
    "org.freedesktop.udisks2.filesystem-mount": YES,
    "org.freedesktop.udisks2.encrypted-unlock": YES,
    "org.freedesktop.udisks2.eject-media": YES,
    "org.freedesktop.udisks2.power-off-drive": YES,
    // only required if starting udiskie via systemd and using udisks2:
    "org.freedesktop.udisks2.filesystem-mount-other-seat": YES,
    "org.freedesktop.udisks2.encrypted-unlock-other-seat": YES,
    "org.freedesktop.udisks2.eject-media-other-seat": YES,
    "org.freedesktop.udisks2.power-off-drive-other-seat": YES
  };

  if (subject.isInGroup("storage")) {
    return permission[action.id];
  }
});
```

**Comment by cherti** [comment]
I thank you very much, sir, for ye help, for ye are very awesome! :) Now it mounts like a charm.

I suspect missing polkit-permissions have been the problem. might there be a possibility to include this polkit-rules-file into the repository so that it gets packaged with udiskie with the next release?

**Comment by coldfix** [comment]
Happy to help, especially bug report quality is as good as yours;).

I don't really think it makes sense to include this in the udiskie repository. Some distributions still use PolicyKit instead of polkit (with .conf style rules), which means it should be the packager's responsibility to include this kind of configuration, IMO. 

Furthermore, I'm not even sure the [arch package](https://www.archlinux.org/packages/community/any/udiskie/) should include this set of rules as it won't be necessary for most users. IMO, it hurts less to manually setup a file into `/etc` than accidentally introducing any kind of security problem by packaging too permissive configuration - KISS. (But you can of course ask the current maintainer.)

I am open to discussion if you have convincing arguments, however. I will improve the documentation and add a remark that the `other-seat` permission is also necessary for this specific case (which I wasn't aware of either).

**Comment by cherti** [comment]
You got a point with the polkit-file above probably including too many options for most people.

But I think the polkit-configuration is worth an article in the wiki. I'd say the polkit file above is more or less sufficient but it should be presented in a place where it is easy to find. If with time it turns out there are more options for special usecases, maybe one can add them too, so that one has an example-polkit-file with all the options that might be relevant in it and the user can then decide which ones are needed and which ones are not.

**Comment by coldfix** [comment]
> ..but it should be presented in a place where it is easy to find

By the way, that very polkit rule is already right on the [front page](https://github.com/coldfix/udiskie#permissions);). But yeah, the current documentation should probably be less clustered and more complete.

**Comment by cherti** [comment]
ok, that now is… a little embarrasing ^^ nothing said.
